File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class CircularBuffer<T>
{
	private readonly int _capacity;

	public CircularBuffer(int capacity) =>
		_capacity = capacity;

	private Dictionary<DateTime, T> _tByTimestamp = new();

	private DateTime _earliestKey =>
		_tByTimestamp.Keys.Min();

	private bool _isFull =>
		_tByTimestamp.Count >= _capacity;

	public T Read()
	{
		if (!_tByTimestamp.Any())
		{
			throw new InvalidOperationException("No data.");
		}
		else
		{
			DateTime earliestTimestamp = _earliestKey;
			T earliestValue = _tByTimestamp[earliestTimestamp];
			_ = _tByTimestamp.Remove(earliestTimestamp);
			return earliestValue;
		}
	}

	public void Write(T value) =>
		_tByTimestamp[DateTime.UtcNow] = _isFull
			? throw new InvalidOperationException("Full.")
			: value;

	public void Overwrite(T value)
	{
		if (_isFull)
		{
			_ = _tByTimestamp.Remove(_earliestKey);
		}

		Write(value);
	}

	public void Clear() =>
		_tByTimestamp = new();
}

File: solutions/scala/circular-buffer/1/src/main/scala/CircularBuffer.scala
class EmptyBufferException
    extends Exception {}

class FullBufferException
    extends Exception {}

class CircularBuffer(var capacity: Int):
    private val data: Array[Int] = new Array[Int](capacity)
    private var readIndex: Int = 0
    private var writeIndex: Int = 0
    private var currentSize: Int = 0

    def write(value: Int): Unit = {
        if (currentSize == capacity)
            throw new FullBufferException()

        data(writeIndex) = value
        writeIndex = (writeIndex + 1) % capacity
        currentSize += 1
    }

    def read(): Int =
        if (currentSize == 0)
            throw new EmptyBufferException()

        val value = data(readIndex)
        readIndex = (readIndex + 1) % capacity
        currentSize -= 1
        value

    def overwrite(value: Int): Unit =
        if currentSize == capacity
        then writeAndAdvance(value)
        else write(value)

    private def writeAndAdvance(value: Int): Unit =
        data(writeIndex) = value
        writeIndex = (writeIndex + 1) % capacity
        readIndex = writeIndex

    def clear(): Unit =
        readIndex = 0
        writeIndex = 0
        currentSize = 0
